### PR TITLE
Disputable Voting: Support delayed execution

### PR DESCRIPTION
**Pull request by facuspagnuolo**
**MERGE AFTER #1195**

### CODE REVIEW

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 1]
```diff
@@ -560,6 +588,16 @@ contract DisputableVoting is DisputableAragonApp, IForwarder {
         emit ChangeOverruleWindow(_overruleWindow);
     }
 
+    /**
+    * @dev Internal function to change the execution delay
+    * @param _executionDelay New execution delay in seconds
+    */
+    function _changeExecutionDelay(uint64 _executionDelay) internal {
+        require(_executionDelay <= voteTime, ERROR_INVALID_EXECUTION_DELAY);
```
Comment (bingen) [inline apps/voting-disputable/contracts/DisputableVoting.sol:596]: What’s the reason for this requirement?

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 2]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
-        it('cannot be executed twice', async () => {
+        it('reverts', async () => {
+          ({ voteId } = await createVote({ voting, script, from }))
+
           await voting.vote(voteId, true, { from: holder51 })
           await voting.mockIncreaseTime(VOTE_DURATION)
-          await voting.executeVote(voteId)
 
-          await assertRevert(voting.executeVote(voteId), VOTING_ERRORS.VOTING_CANNOT_EXECUTE)
+          await assertRevert(voting.executeVote(voteId), 'EVMCALLS_INVALID_LENGTH')
```
Comment (bingen) [inline apps/voting-disputable/test/contracts/voting_execution.js:121]: Can we add this one to the error constants file?

File: apps/voting-disputable/test/contracts/voting_settings.js [hunk 3]
```diff
@@ -209,4 +210,54 @@ contract('Voting settings', ([_, owner, anyone, holder51, holder20, holder29]) =
       })
     })
   })
+
+  describe('changeExecutionDelay', () => {
+    context('when the sender is allowed', () => {
+      const from = owner
+
+      context('when the new window is valid', () => {
+        const newDelay = DAY
```
Comment (bingen) [inline apps/voting-disputable/test/contracts/voting_settings.js:219]: This is the same as the old one, right? We should add test for real changes of the value.

File: apps/voting-disputable/test/contracts/voting_settings.js [hunk 4]
```diff
@@ -209,4 +210,54 @@ contract('Voting settings', ([_, owner, anyone, holder51, holder20, holder29]) =
...
+        })
+
+        it('emits an event', async () => {
+          const receipt = await voting.changeExecutionDelay(newDelay, { from })
+
+          assertAmountOfEvents(receipt, 'ChangeExecutionDelay')
+          assertEvent(receipt, 'ChangeExecutionDelay', { executionDelay: newDelay })
+        })
+
+        it('does not affect previous created votes', async () => {
```
Comment (bingen) [inline apps/voting-disputable/test/contracts/voting_settings.js:234]: Well, the value didn’t actually changed ;-)

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 5]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
+    let voteId, executionTarget, script
 
-    context('with an empty script', () => {
-      beforeEach('create script', async () => {
-        ({ script } = await voteScript(0))
-      })
-
-      it('cannot be executed', async () => {
-        ({ voteId } = await createVote({ voting, script, from }))
+    context('without delayed execution', () => {
```
Comment (bingen) [inline apps/voting-disputable/test/contracts/voting_execution.js:29]: Seems like we could share a lot of the tests between the case of delayed execution = 0 and the one > 0.

File: apps/voting-disputable/test/contracts/voting_execution.js (hunk 2 above)
Comment (facuspagnuolo) [reply apps/voting-disputable/test/contracts/voting_execution.js:121]: Added in https://github.com/aragon/aragon-apps/pull/1195/commits/3b0d3a0d4cb74db5d6af759e10237cffa3d4d3d2

File: apps/voting-disputable/test/contracts/voting_settings.js (hunk 3 above)
Comment (facuspagnuolo) [reply apps/voting-disputable/test/contracts/voting_settings.js:219]: Yea! Good catch!

File: apps/voting-disputable/contracts/DisputableVoting.sol (hunk 1 above)
Comment (facuspagnuolo) [reply apps/voting-disputable/contracts/DisputableVoting.sol:596]: Simply so it cannot be extended forever, I'll double check with @lkngtn and @john-light

Comment (facuspagnuolo) [reply apps/voting-disputable/contracts/DisputableVoting.sol:596]: Checked with @lkngtn, let's remove the validation and enforce this via frontend

File: apps/voting-disputable/contracts/DisputableVoting.sol [hunk 6]
```diff
@@ -28,6 +28,9 @@ contract DisputableVoting is DisputableAragonApp, IForwarder {
     // bytes32 public constant MODIFY_OVERRULE_WINDOW_ROLE = keccak256("MODIFY_OVERRULE_WINDOW_ROLE");
     bytes32 public constant MODIFY_OVERRULE_WINDOW_ROLE = 0x703200758fea823fd0c9d36774ba318b14f6c50c4ef6dff34f394e8f2c7d2d99;
 
+    // bytes32 public constant MODIFY_EXECUTION_DELAY_ROLE = keccak256("MODIFY_EXECUTION_DELAY_ROLE");
+    bytes32 public constant MODIFY_EXECUTION_DELAY_ROLE = 0x69a0bddd05e66b7ae81cce9994caef3b5c660de549fd2fd3597a9e2c3046e446;
```
Comment (sohkai) [inline apps/voting-disputable/contracts/DisputableVoting.sol:32]: Don't forget to put this into the arapp.json file :)

File: apps/voting-disputable/contracts/DisputableVoting.sol (hunk 1 above)
Comment (sohkai) [inline apps/voting-disputable/contracts/DisputableVoting.sol:596]: Not sure if this condition should exist; theoretically it shouldn't matter how long your execution delay is for.

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 7]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
+          })
+
+          it('cannot execute vote if not enough quorum met', async () => {
+            await voting.vote(voteId, true, { from: holder20 })
+            await voting.mockIncreaseTime(VOTE_DURATION)
+
+            await assertRevert(voting.executeVote(voteId), VOTING_ERRORS.VOTING_CANNOT_EXECUTE)
+          })
+
+          it('cannot execute vote if not support met', async () => {
```
Comment (sohkai) [inline apps/voting-disputable/test/contracts/voting_execution.js:73]:
```suggestion
          it('cannot execute vote if not enough support met', async () => {
```

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 8]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
-        it('cannot execute vote if not support met', async () => {
-          await voting.vote(voteId, false, { from: holder29 })
-          await voting.vote(voteId, false, { from: holder20 })
-          await voting.mockIncreaseTime(VOTE_DURATION)
+          it('is not automatically executed', async () => {
+            assertBn(await executionTarget.counter(), 0, 'should have received execution call')
+          })
 
-          await assertRevert(voting.executeVote(voteId), VOTING_ERRORS.VOTING_CANNOT_EXECUTE)
+          it('cannot be executed immediately executed', async () => {
```
Comment (sohkai) [inline apps/voting-disputable/test/contracts/voting_execution.js:62]:
```suggestion
          it('cannot be immediately executed', async () => {
```

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 9]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
-      it('reverts', async () => {
+      beforeEach('create vote', async () => {
         ({ voteId } = await createVote({ voting, script, from }))
+      })
+
+      it('is not automatically executed', async () => {
+        assertBn(await executionTarget.counter(), 0, 'should have received execution call')
+      })
+
+      it('cannot be executed immediately executed', async () => {
```
Comment (sohkai) [inline apps/voting-disputable/test/contracts/voting_execution.js:145]:
```suggestion
      it('cannot be immediately executed', async () => {
```

File: apps/voting-disputable/test/contracts/voting_execution.js [hunk 10]
```diff
@@ -22,98 +22,168 @@ contract('Voting', ([_, owner, holder20, holder29, holder51]) => {
...
+      })
+
+      it('cannot execute vote if not enough quorum met', async () => {
+        await voting.vote(voteId, true, { from: holder20 })
+        await voting.mockIncreaseTime(VOTE_DURATION + EXECUTION_DELAY)
+
+        await assertRevert(voting.executeVote(voteId), VOTING_ERRORS.VOTING_CANNOT_EXECUTE)
+      })
+
+      it('cannot execute vote if not support met', async () => {
```
Comment (sohkai) [inline apps/voting-disputable/test/contracts/voting_execution.js:156]:
```suggestion
      it('cannot execute vote if not enough support met', async () => {
```

**Pull request merged**